File: 234 Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    
    public ListNode reverseLinkedlist(ListNode head){
        ListNode current = head;
        ListNode next = null;
        ListNode prev = null;
        
        while(current != null){
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        
        return prev;
    }
    
    public ListNode middle(ListNode head){
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        
        return slow;
    }
    
    public boolean isPalindrome(ListNode head) {
        ListNode middle = middle(head);
        ListNode reversedSecondHalf = reverseLinkedlist(middle);
        
        while(reversedSecondHalf != null){
            if(reversedSecondHalf.val != head.val){
                return false;
            }
            reversedSecondHalf = reversedSecondHalf.next;
            head = head.next;
        }
        
        return true;
    }
}


// 84 / 84 test cases passed.
// Status: Accepted
// Runtime: 6 ms
// Memory Usage: 49.5 MB

File: 234 Palindrome Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* reverse(struct ListNode* head){
    if(head == NULL)
        return NULL;
    struct ListNode* curr = head;
    struct ListNode* prev = NULL;
    struct ListNode* NextNode = NULL;

    while(curr){
        NextNode = curr -> next;
        curr-> next = prev;
        prev = curr;
        curr = NextNode;
    }
    return prev;
}

bool isPalindrome(struct ListNode* head) {

    if(head == NULL)
        return true;
    struct ListNode* fast = head;
    struct ListNode* slow = head; 

    while(fast -> next != NULL && fast -> next -> next != NULL){
        slow = slow -> next;
        fast = fast -> next ->next;
    }
    slow -> next = reverse(slow -> next);
    slow = slow -> next;

    while(slow != NULL){
        if(head -> val != slow -> val)
            return false;
        slow = slow -> next;
        head = head -> next;
    }
    return true;
}
// Runtime: 128 ms, faster than 26.09% of C online submissions for Palindrome Linked List.
// Memory Usage: 41.1 MB, less than 53.23% of C online submissions for Palindrome Linked List.
